src/System.IO.Pipelines/Pipe.cs

```diff
@@ -592,9 +657,14 @@ private void Dispose()
         void IReadableBufferAwaiter.OnCompleted(Action continuation)
         {
             Action awaitable;
+            bool doubleCompletion;
             lock (_sync)
             {
-                awaitable = _readerAwaitable.OnCompleted(continuation, ref _writerCompletion);
+                awaitable = _readerAwaitable.OnCompleted(continuation, out doubleCompletion);
+            }
+            if (doubleCompletion)
+            {
+                Writer.Complete(PipelinesThrowHelper.GetInvalidOperationException(ExceptionResource.NoConcurrentOperation));
```

Comment:
AwaitingFlushAsyncAwaitableTwiceCompletesReaderWithException checks both tasks for `            Assert.Equal(true, task2.IsCompleted);
`